### PR TITLE
Add RetryStatus enum and use it

### PR DESCRIPTION
**Pull request by alexshtin**
Adding `RetryStatus` enum and field of this type to `Failure`. Generally, it indicates the reason why server stopped retrying (can be useful for debugging). There is also some extra logic on server based on this field.

### CODE REVIEW

File: event/message.proto [hunk 1]
```diff
@@ -191,13 +191,10 @@ message ActivityTaskFailedEventAttributes {
...
 message ActivityTaskTimedOutEventAttributes {
-    common.Payloads lastHeartbeatDetails = 1;
     int64 scheduledEventId = 2;
     int64 startedEventId = 3;
-    common.TimeoutType timeoutType = 4;
-    // For retry activity, it may have a failure before timeout. It's important to keep those information for debug.
-    // Client can also provide the info for making next decision
-    failure.Failure lastFailure = 5;
+    // For retry activity, it may have a failure before timeout. It is stored as cause in timeoutFailure.
+    failure.Failure timeoutFailure = 5;
```
Comment (alexshtin) [inline event/message.proto:197]: This always will be failure with `TimeoutFailureInfo`. It has `lastHeartbeatDetails` and `timeoutType`. `lastFailure` will go to `cause`.

Comment (mfateev) [reply event/message.proto:197]: It is not clear. What is top level failure returned?

Comment (alexshtin) [reply event/message.proto:197]: Top level is always `TimeoutFailureInfo` (I guess, with `ScheduleToClose` type always). `cause` might be another `TimeoutFailureInfo` with `StartToClose` or something else from `mutableState`.

File: common/enum.proto [hunk 2]
```diff
@@ -70,3 +70,13 @@ enum TimeoutType {
...
 }
+
+enum RetryStatus {
+    Retrying = 0; // or InProgress, or Active
+    NonRetryableError = 1;
+    Timeout = 2;
+    MaximumAttemptsExceeded = 3; // or MaximumAttemptsExhausted
+    Disabled = 4; // or RetryDisabled, or NotConfigured
+    InternalServerError = 5; // InternalError, ServiceError
+    CancelRequested = 6;
```
Comment (alexshtin) [inline common/enum.proto:81]: Need some input on these names.

File: failure/message.proto [hunk 3]
```diff
@@ -88,4 +87,5 @@ message Failure {
         ActivityTaskFailureInfo activityTaskFailureInfo = 11;
         ChildWorkflowExecutionFailureInfo childWorkflowExecutionFailureInfo = 12;
     }
+    common.RetryStatus retryStatus = 13;
```
Comment (mfateev) [inline failure/message.proto:90]: This doesn't belong to every error. It should be part of ActivityTaskFailureInfo and  ChildWorkflowExecutionFailureInfo.

File: event/message.proto [hunk 4]
```diff
@@ -381,6 +382,7 @@ message ChildWorkflowExecutionFailedEventAttributes {
     common.WorkflowType workflowType = 4;
     int64 initiatedEventId = 5;
     int64 startedEventId = 6;
+    common.RetryStatus retryStatus = 7;
```
Comment (mfateev) [inline event/message.proto:385]: We never retry cancelled activities. So I dont think it should be here.

File: event/message.proto [hunk 5]
```diff
@@ -399,6 +401,7 @@ message ChildWorkflowExecutionTimedOutEventAttributes {
...
```
Comment (mfateev) [inline event/message.proto:404]: Cancelled children are never retried.

**Pull request merged**